### DISCUSSION TITLE
Add support for group membership claims

### DISCUSSION THREAD

**Discussion by MrMossevig** [Azure/data-api-builder#1855] [Ideas]
Hi,

We're using the new Database Connection with SWA. We also want to use the built-in Authentication with Entra ID of the SWA to keep things simple and avoid building glue ourselves.

We're using security groups to control who has access to what data. I.e. a security group has access to a project and it's corresponding data in the database. If I understand the API builder correctly, it only supports roles today and not group claims?

Are there any plans for implementing it or other ways to acheive  the same thing?

Ref: This is the conversation that started this Feature Request: https://learn.microsoft.com/en-us/answers/questions/1410852/does-the-new-database-connection-in-static-web-app

**Comment by MrMossevig** [comment]
To add an example to the problem, let's say you have the entity Books, and the roles User and Author. The User can read all books, and the Author can also edit her own, but only her own, Books. How do you acheive that in the API Builder?

**Reply by yorek** [reply]
You need to use "database policies" or a database, like Azure SQL or SQL Server, that supports natively row-level security.

With database policies you can limit what someone can do based on the claims they have. For example a policy like

```
@claims.user_id eq @item.owner_id
```
would allow a user only to access the book for which the value of the claim `user_id` is equal to the value of the column or property `owner_id` in the related database object.

If you need something more complex, with the `mssql` driver, you can pass *all* the claims to the database engine, and then use Row-Level Security to implement any row-level access logic you want: [Row-level security](https://learn.microsoft.com/en-us/sql/relational-databases/security/row-level-security?view=sql-server-ver16)

**Reply by MrMossevig** [reply]
Okay, that's quite cool. Do you know where to find more information about how to set the claims in SWA claims and pass it on to database through db-connection?
What we want to acheive is that we have several Groups in AD (Entra ID) which are mirrored in a database table called organizations. Each member of these organizations/Groups should only be able to access the objects where the Organization have access.

**Reply by MrMossevig** [reply]
@yorek So I've come some way. I guess the policy you're referring to are these: https://learn.microsoft.com/en-us/azure/data-api-builder/configuration-file#policies ?

Do you know how to use the groups provided in claims-array to limit access?

**Reply by seantleonard** [reply]
DAB doesn't support the groups claim for authZ but uses roles. We have that on our radar for supporting. There is also a limitation of SWA that it does not pass all claims to DAB and than is documented here:
-  https://learn.microsoft.com/en-us/azure/static-web-apps/user-information?tabs=csharp#api-functions:~:text=Note,the%20claims%20array. 
    - Source: https://github.com/Azure/static-web-apps/issues/897#issuecomment-1904238486

**Reply by MrMossevig** [reply]
Thank you! This might be drifting into off-topic territory, but it might help us solve the case. Is it possible to use related fields in the policies? E.g. If I have a many-to-one relationship between books and organizations and each organization have a an id, can I then  use @item.organization.id  in the policy?

**Reply by seantleonard** [reply]
Unfortunately, no. This is being tracked by https://github.com/Azure/data-api-builder/issues/569

**Reply by MrMossevig** [reply]
Okay, got it. And I guess the answer to my last question is also no, using part of the userDetails (e.g. the domain (split at '@')) for filtering?

**Reply by seantleonard** [reply]
DAB can't parse out a substring of `userDetails`, however I wouldn't recommend using that field alone for authorization purposes since the username is mutable. Using the userId would probably be better in that case.

**Reply by MrMossevig** [reply]
Yeah, it was a bit too late when I thought about that. Agree - not a good idea :)

**Comment by MrMossevig** [comment]
@JerryNixon Was this resolved? I didn't see it in any of the releases, but I might have missed it?
